File: server/src/main/java/com/willkamp/vial/implementation/Route.java
package com.willkamp.vial.implementation;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Borrowed from @see <href= https://github.com/xjdr/xio>xio</href=> */
class Route {

  private static final Pattern keywordPattern = Pattern.compile("(:\\w+|:\\*\\w+)");
  private final String path;
  private final Pattern pathPattern;
  private final List<String> keywords;

  private Route(String path, Pattern pathPattern, List<String> keywords) {
    this.path = path;
    this.pathPattern = pathPattern;
    this.keywords = keywords;
  }

  private static Pattern compile(String pattern, List<String> keywords) {
    StringBuilder regexPattern = new StringBuilder();

    if (pattern.equals("/")) {
      regexPattern.append("/");
    } else {
      final String[] segments = pattern.split("/");

      for (String segment : segments) {
        if (!segment.equals("")) {
          regexPattern.append("/");
          if (keywordPattern.matcher(segment).matches()) {
            String keyword = segment.substring(1);

            if (keyword.indexOf("*") == 0) {
              keyword = keyword.substring(1);
              regexPattern.append("(?<").append(keyword).append(">.*)");
            } else {
              regexPattern.append("(?<").append(keyword).append(">[^/]*)");
            }
            keywords.add(keyword);
          } else {
            regexPattern.append(segment);
          }
        }
      }
    }
    regexPattern.append("[/]?");

    return Pattern.compile(regexPattern.toString());
  }

  static Route build(String pattern) {
    List<String> keywords = new ArrayList<>();
    return new Route(pattern, compile(pattern, keywords), keywords);
  }

  Pattern pathPattern() {
    return pathPattern;
  }

  boolean matches(CharSequence path) {
    return pathPattern.matcher(path).matches();
  }

  Map<String, String> groups(CharSequence path) {
    Matcher matcher = pathPattern.matcher(path);
    Map<String, String> groups = new HashMap<>();
    if (matcher.matches()) {
      for (String keyword : keywords) {
        groups.put(keyword, matcher.group(keyword));
      }
    }
    return groups;
  }

  // Let's just assume that if two Route objects have been built
  // from the same path that they will have the same pattern and
  // keywords.

  @Override
  public boolean equals(final Object o) {
    if (o == this) return true;
    if (!(o instanceof Route)) return false;
    final Route other = (Route) o;
    return other.path.equals(path);
  }

  @Override
  public int hashCode() {
    return path.hashCode();
  }
}

File: server/src/main/java/com/willkamp/vial/implementation/Route.kt
package com.willkamp.vial.implementation

import java.util.ArrayList
import java.util.HashMap
import java.util.regex.Pattern

internal class Route private constructor(private val path: String, private val pathPattern: Pattern, private val keywords: List<String>) {

    fun pathPattern(): Pattern {
        return pathPattern
    }

    fun matches(path: CharSequence): Boolean {
        return pathPattern.matcher(path).matches()
    }

    fun groups(path: CharSequence): Map<String, String> {
        val matcher = pathPattern.matcher(path)
        val groups = HashMap<String, String>()
        if (matcher.matches()) {
            for (keyword in keywords) {
                groups[keyword] = matcher.group(keyword)
            }
        }
        return groups
    }

    // Let's just assume that if two Route objects have been built
    // from the same path that they will have the same pattern and
    // keywords.

    override fun equals(other: Any?): Boolean {
        return if (other === this) {
            true
        } else {
            other?.takeIf {
                it is Route && it.path == path
            }?.let {
               true
            }?: false
        }
    }

    override fun hashCode(): Int {
        return path.hashCode()
    }

    companion object {

        private val keywordPattern = Pattern.compile("(:\\w+|:\\*\\w+)")

        private fun compile(pattern: String, keywords: MutableList<String>): Pattern {
            val regexPattern = StringBuilder()

            if (pattern == "/") {
                regexPattern.append("/")
            } else {
                val segments = pattern.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                for (segment in segments) {
                    if (segment != "") {
                        regexPattern.append("/")
                        if (keywordPattern.matcher(segment).matches()) {
                            var keyword = segment.substring(1)

                            if (keyword.indexOf("*") == 0) {
                                keyword = keyword.substring(1)
                                regexPattern.append("(?<").append(keyword).append(">.*)")
                            } else {
                                regexPattern.append("(?<").append(keyword).append(">[^/]*)")
                            }
                            keywords.add(keyword)
                        } else {
                            regexPattern.append(segment)
                        }
                    }
                }
            }
            regexPattern.append("[/]?")

            return Pattern.compile(regexPattern.toString())
        }

        fun build(pattern: String): Route {
            val keywords = ArrayList<String>()
            return Route(pattern, compile(pattern, keywords), keywords)
        }
    }
}
